src/dotty/tools/dotc/transform/TypeTestsCasts.scala

```diff
@@ -90,10 +91,12 @@ trait TypeTestsCasts {
           }
           else if (argCls.isPrimitiveValueClass)
             unbox(qual.ensureConforms(defn.ObjectType), argType)
-          else
+          else if (isDerivedValueClass(argCls)) {
+            qual // adaptToType in Erasure will do the necessary type adaptation
+          } else
             derivedTree(qual, defn.Any_asInstanceOf, argType)
         }
-        def erasedArg = erasure(tree.args.head.tpe)
+        def erasedArg = erasure(tree.args.head.tpe, isSemi = false)
```

Comment:
Tests like `isInstanceOf[Meter]` and `asInstanceOf|Meter]` should not be erased to `isInstanceOf[ErasedValueType(Meter, ...)]`. `IsInstanceOf` tests with value classes are like tests with normal value classes, `asInstanceOf` is special see `else if (isDerivedValueClass(argCls)) {`